### PR TITLE
Bump new packages

### PR DESCRIPTION
**Pull request by sunnamed434**
NB! Breaking changes!

## NuGet packages changes:
- `Microsoft.Extensions.* v3.1.32` -> `Microsoft.Extensions.* v7.*, etc`

- ~~`Pomelo.EntityFrameworkCore.MySql v3.2.7` -> `Pomelo.EntityFrameworkCore.MySql v5.0.4`~~ 
Decided to use `Pomelo.EntityFrameworkCore.MySql v3.2.7` as before.

- ~~`Microsoft.EntityFrameworkCore. v.3.1.32` -> `Microsoft.EntityFrameworkCore.* v.5.`~~
Decided to use `Microsoft.EntityFrameworkCore. v5.0. 14`.

- `Lib.Harmony v2.2.2` -> `Lib.Harmony v2.3-prerelease.3`

Now OpenMod uses `netstandard2.1` and `netstandard2.0` only, don't use/support any more net framework

## Tested with
- uScript installed (it's ok)
- uScript Plugins:
  - Soon...
- Harmony (it's ok)
- RocketMod plugins:
  - `UnturnedGuard` (work fine)
  - [ShimmysAdminTools](https://github.com/ShimmyMySherbet/ShimmysAdminTools) (workfine)
- OpenMod plugins:
  - `Pustalorc.PlayerInfoLib.Unturned.3.6.6` (work fine)
  - `DiFFoZ.Kits` (work fine)
  - `Wild.Announcer` (work fine)
  - `New Essentials` (work fine)
  - `OpenMod.Economy` (work fine)
  - `UnturnedGuard` (work fine)

## Deprecated plugins
Plugins that use the provided dependencies will be deprecated:
- ~~ILogger (because ILogger moved into another package: `Microsoft.Extensions.Logging.Abstractions v7.0.1`)~~
it's ok now
- ServerVersion methods or options.ServerVersion
```cs
options.UseMySql(connectionString, sqlOptions =>
{
    // other options configuration...
    sqlOptions.ServerVersion(serverVersion); // This line should be removed now.
});
```

## TL;DR
It's not done yet, still in the process of testing and development, sending it here to hear feedback and fix problems earlier

Yes, a lot of plugins will deprecate, but we will gonna need to proceed with a bit of work because when Unity 2022 is used in Unturned it will cause us a lot more trouble than now - better now than in the last moment.

### CODE REVIEW

**Comment by sunnamed434** [comment]
@dotnet-policy-service agree

**Comment by rube200** [comment]
what about Microsoft.Extensions.* 7.*?

**Comment by sunnamed434** [comment]
> what about Microsoft.Extensions.* 7.*?

We can't use it while Unity on NET Framework and Netstandard, the Extensions starting with 6 and higher has target framework .NET 6/7/8 only, etc

we will be tied for a long time with netstandard2.1

File: framework/OpenMod.Core/Localization/NullStringLocalizer.cs [hunk 1]
```diff
@@ -24,10 +24,6 @@ public IEnumerable<LocalizedString> GetAllStrings(bool includeParentCultures)
             return Enumerable.Empty<LocalizedString>();
         }
 
-        public IStringLocalizer WithCulture(CultureInfo culture)
-        {
-            return this;
-        }
```
Comment (rube200) [inline framework/OpenMod.Core/Localization/NullStringLocalizer.cs:30]: Maybe instead of removing, mark it as obsolete and just return IStringLocalizer

File: framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs [hunk 2]
```diff
@@ -21,13 +21,6 @@ public IEnumerable<LocalizedString> GetAllStrings(bool includeParentCultures)
             return m_StringLocalizer.GetAllStrings(includeParentCultures);
         }
 
-#pragma warning disable 618 // disable obsolete warning
-        public IStringLocalizer WithCulture(CultureInfo culture)
-        {
-            return m_StringLocalizer.WithCulture(culture);
-        }
-#pragma warning restore 618
-
```
Comment (rube200) [inline framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs:30]: Maybe instead of removing, mark it as obsolete and just return IStringLocalizer

**Comment by rube200** [comment]
> > what about Microsoft.Extensions.* 7.*?
> 
> We can't use it while Unity on NET Framework and Netstandard, the Extensions starting with 6 and higher has target framework .NET 6/7/8 only, etc
> 
> we will be tied for a long time with netstandard2.1

I just checked and nuget and seems like they are also netstandard2.0

File: rust/OpenMod.Rust/OpenMod.Rust.csproj [hunk 3]
```diff
@@ -12,8 +12,8 @@
     <PackageReference Include="OpenMod.Rust.Redist" Version="2260.199.1">
       <IncludeAssets>compile</IncludeAssets>
     </PackageReference>
-    <PackageReference Include="OpenMod.UniTask" Version="2019.2.2.5" />
-    <PackageReference Include="OpenMod.UnityEngine.Redist" Version="2019.4.10">
+    <PackageReference Include="OpenMod.UniTask" Version="2021.2.4.1" />
+    <PackageReference Include="OpenMod.UnityEngine.Redist" Version="2021.3.28.1">
```
Comment (rube200) [inline rust/OpenMod.Rust/OpenMod.Rust.csproj:16]: Did you check Unity version of Rust before changing this?

Comment (sunnamed434) [reply rust/OpenMod.Rust/OpenMod.Rust.csproj:16]: yeah, Rust version is 2022 now

Comment (sunnamed434) [reply rust/OpenMod.Rust/OpenMod.Rust.csproj:16]: but I can be wrong, I just saw somewhere that they have now 2022 Unity Engine

**Comment by sunnamed434** [comment]
> > > what about Microsoft.Extensions.* 7.*?
> > 
> > 
> > We can't use it while Unity on NET Framework and Netstandard, the Extensions starting with 6 and higher has target framework .NET 6/7/8 only, etc
> > we will be tied for a long time with netstandard2.1
> 
> I just checked and nuget and seems like they are also netstandard2.0

Pomelo.EntityFrameworkCore.MySql will be a problem then
image

File: framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs (hunk 2 above)
Comment (sunnamed434) [reply framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs]: done

File: framework/OpenMod.Core/Localization/NullStringLocalizer.cs (hunk 1 above)
Comment (sunnamed434) [reply framework/OpenMod.Core/Localization/NullStringLocalizer.cs]: done

**Comment by sunnamed434** [comment]
I guess its time to remove [this](https://github.com/openmod/openmod/blob/main/props/SharedProjectProps.props#L46-L64) also

File: extensions/OpenMod.Extensions.Economy.Abstractions/OpenMod.Extensions.Economy.Abstractions.csproj [hunk 4]
```diff
@@ -1,7 +1,7 @@
 <Project Sdk="Microsoft.NET.Sdk">
```
Comment (CyberAndrii) [inline extensions/OpenMod.Extensions.Economy.Abstractions/OpenMod.Extensions.Economy.Abstractions.csproj:1]: Also update target framework in https://github.com/openmod/openmod/tree/main/templates and https://github.com/openmod/openmod/tree/main/tools

**Review by CyberAndrii — CHANGES_REQUESTED** [review]
> because when Unity 2022 is used in Unturned it will cause us a lot more trouble than now

Why?

> > > > what about Microsoft.Extensions.* 7.*?
> > > 
> > > 
> > > We can't use it while Unity on NET Framework and Netstandard, the Extensions starting with 6 and higher has target framework .NET 6/7/8 only, etc
> > > we will be tied for a long time with netstandard2.1
> > 
> > 
> > I just checked and nuget and seems like they are also netstandard2.0
> 
> Pomelo.EntityFrameworkCore.MySql will be a problem then

I'm still confused why is that an issue. Is there an exception or something? Don't update Pomelo.EntityFrameworkCore.MySql but just Microsoft.Extensions.

>   - `Pustalorc.PlayerInfoLib.Unturned.3.6.6` (doesn't work anymore)
>   - `DiFFoZ.Kits` (doesn't work anymore)
>   - `Wild.Announcer` (doesn't work anymore)

I don't think we can just break them. Maybe there are some workarounds.

File: framework/OpenMod.Core/Users/OfflineUserProvider.cs [hunk 5]
```diff
@@ -76,7 +76,7 @@ public async Task<bool> BanAsync(IUser user, IUser? instigator = null, string? r
 
             expireDate ??= DateTime.MaxValue;
             reason ??= m_StringLocalizer["ban_default"];
-            data.BanInfo = new BanData(reason, instigator, expireDate);
+            data.BanInfo = new BanData(reason!, instigator, expireDate);
```
[the remark this answers is not in the document]
Comment (rube200) [reply framework/OpenMod.Core/Users/OfflineUserProvider.cs:79]: What happens when IStringLocalizer can't find the string by name?

[the remark this answers is not in the document]
Comment (sunnamed434) [reply framework/OpenMod.Core/Users/OfflineUserProvider.cs:79]: > What happens when IStringLocalizer can't find the string by name?

it will just return an inputted value, in this case: `ban_default`

File: framework/OpenMod.Core/Commands/CommandStoreOptions.cs [hunk 6]
```diff
@@ -37,7 +37,7 @@ public IReadOnlyCollection<ICommandSource> CreateCommandSources(IServiceProvider
 
             foreach (var type in m_CommandSourceTypes)
             {
-                sources.Add((ICommandSource)ActivatorUtilitiesEx.CreateInstance(lifetime, type));
+                sources.Add((ICommandSource)ActivatorUtilitiesEx.CreateInstance(lifetime!, type));
```
[the remark this answers is not in the document]
Comment (sunnamed434) [reply framework/OpenMod.Core/Commands/CommandStoreOptions.cs:40]: Done

File: framework/OpenMod.Analyzers/OpenMod.Analyzers.csproj [hunk 7]
```diff
@@ -3,7 +3,7 @@
   <PropertyGroup>
     <PackageDescription>CSharp Analyzers for OpenMod.</PackageDescription>
     <MinClientVersion>3.6</MinClientVersion>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (sunnamed434) [reply framework/OpenMod.Analyzers/OpenMod.Analyzers.csproj]: Done

File: unturned/OpenMod.Unturned/Users/UnturnedUserProvider.cs [hunk 8]
```diff
@@ -356,7 +356,7 @@ public async Task<bool> BanAsync(IUser user, IUser? instigator = null, string? r
             var data = await m_UserDataStore.GetUserDataAsync(user.Id, user.Type);
             if (data != null)
             {
-                data.BanInfo = new BanData(reason, instigator, expireDate);
+                data.BanInfo = new BanData(reason!, instigator, expireDate);
```
[the remark this answers is not in the document]
Comment (sunnamed434) [reply unturned/OpenMod.Unturned/Users/UnturnedUserProvider.cs:359]: Now reason is nullable

File: rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj [hunk 9]
```diff
@@ -1,6 +1,6 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFramework>net461</TargetFramework>
+    <TargetFramework>netstandard2.1</TargetFramework>
```
[the remark this answers is not in the document]
Comment (sunnamed434) [reply rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj:3]: If I make it net standard then I'll gonna need to update other projects and keep support for net461 (OpenMod. Core/Common/API, etc)

File: extensions/OpenMod.Extensions.Economy.Abstractions/OpenMod.Extensions.Economy.Abstractions.csproj (hunk 4 above)
Comment (sunnamed434) [reply extensions/OpenMod.Extensions.Economy.Abstractions/OpenMod.Extensions.Economy.Abstractions.csproj:1]: Done

**Comment by sunnamed434** [comment]
> > because when Unity 2022 is used in Unturned it will cause us a lot more trouble than now
> 
> Why?
> 
> > > > > what about Microsoft.Extensions.* 7.*?
> > > > 
> > > > 
> > > > We can't use it while Unity on NET Framework and Netstandard, the Extensions starting with 6 and higher has target framework .NET 6/7/8 only, etc
> > > > we will be tied for a long time with netstandard2.1
> > > 
> > > 
> > > I just checked and nuget and seems like they are also netstandard2.0
> > 
> > 
> > Pomelo.EntityFrameworkCore.MySql will be a problem then
> 
> I'm still confused why is that an issue. Is there an exception or something? Don't update Pomelo.EntityFrameworkCore.MySql but just Microsoft.Extensions.
> 
> > * `Pustalorc.PlayerInfoLib.Unturned.3.6.6` (doesn't work anymore)
> > * `DiFFoZ.Kits` (doesn't work anymore)
> > * `Wild.Announcer` (doesn't work anymore)
> 
> I don't think we can just break them. Maybe there are some workarounds.

Now this branch uses Extensions v7.* and old Pomelo

File: framework/OpenMod.API/Users/BanData.cs [hunk 10]
```diff
@@ -19,14 +19,16 @@
...
-        public BanData(string reason, IPermissionActor? instigator = null, DateTime? expireDate = null)
+        public BanData(string? reason = null, IPermissionActor? instigator = null, DateTime? expireDate = null)
         {
             ExpireDate = expireDate ?? DateTime.MaxValue;
 
             InstigatorType = instigator?.Type;
             InstigatorId = instigator?.Id;
 
-            Reason = reason;
+            Reason = string.IsNullOrWhiteSpace(reason) == false
```
[the remark this answers is not in the document]
Comment (sunnamed434) [reply framework/OpenMod.API/Users/BanData.cs]: Done

**Comment by sunnamed434** [comment]
Problem:

```console
[12:56:59 ERR][UnturnedGuard.OpenMod.Logging.OpenModGuardLogger] An unexpected error occured while searching for a feature types!
System.TypeLoadException: Could not load type of field 'UnturnedGuard.Services.ServicesCollectionExtensions+<>c__DisplayClass6_1:serverVersion' (0) due to: Could not resolve type with token 01000027 from typeref (expected class 'Pomelo.EntityFrameworkCore.MySql.Storage.ServerVersion' in assembly 'Pomelo.EntityFrameworkCore.MySql, Version=3.2.7.0, Culture=neutral, PublicKeyToken=2cc498582444921b') assembly:Pomelo.EntityFrameworkCore.MySql, Version=3.2.7.0, Culture=neutral, PublicKeyToken=2cc498582444921b type:Pomelo.EntityFrameworkCore.MySql.Storage.ServerVersion member:(null)
  at (wrapper managed-to-native) System.RuntimeTypeHandle.type_is_assignable_from(System.Type,System.Type)
  at System.RuntimeTypeHandle.CanCastTo (System.RuntimeType type, System.RuntimeType target) [0x00000] in <47fc8c70fa834cbf8141d7c1a7589125>:0
  at System.RuntimeType.IsAssignableFrom (System.Type c) [0x00020] in <47fc8c70fa834cbf8141d7c1a7589125>:0
  at UnturnedGuard.Services.ServicesCollectionExtensions+<>c.<RegisterFeatures>b__1_0 (System.Type t) [0x00000] in <8a4886ce9e5d4379b6cf9f641dc1eb5a>:0
  at System.Linq.Enumerable+WhereArrayIterator`1[TSource].MoveNext () [0x0002c] in <d61464951b0243f18092630834a7bf18>:0
  at System.Collections.Generic.List`1[T].AddEnumerable (System.Collections.Generic.IEnumerable`1[T] enumerable) [0x00059] in <47fc8c70fa834cbf8141d7c1a7589125>:0
  at System.Collections.Generic.List`1[T].InsertRange (System.Int32 index, System.Collections.Generic.IEnumerable`1[T] collection) [0x000f4] in <47fc8c70fa834cbf8141d7c1a7589125>:0
  at System.Collections.Generic.List`1[T].AddRange (System.Collections.Generic.IEnumerable`1[T] collection) [0x00000] in <47fc8c70fa834cbf8141d7c1a7589125>:0
  at UnturnedGuard.Services.ServicesCollectionExtensions.RegisterFeatures (Microsoft.Extensions.DependencyInjection.IServiceCollection source, UnturnedGuard.Guard.Logging.IGuardLogger logger, UnturnedGuard.Shared.Models.UnturnedGuardSettings settings) [0x0004e] in <8a4886ce9e5d4379b6cf9f641dc1eb5a>:0
```

2 Solutions:
- Use v5.* of the Pomelo
- Don't use options.ServerVersion()

As I said before to avoid misunderstanding by other devs using Microsoft.* and Pomelo API it is better to use Pomelo and Ef Core with v5.*

File: framework/OpenMod.Core/Localization/NullStringLocalizer.cs (hunk 1 above)
Comment (DiFFoZ) [reply framework/OpenMod.Core/Localization/NullStringLocalizer.cs]: It should be removed, it's now unused in OpenMod. Obsolete will not work here, due to using `IStringLocalization` interface.

File: framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs (hunk 2 above)
Comment (DiFFoZ) [reply framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs]: It should be removed, it's now unused in OpenMod. Obsolete will not work here, due to using `IStringLocalization` interface.

File: rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj (hunk 9 above)
[the remark this answers is not in the document]
Comment (DiFFoZ) [reply rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj:3]: Use then `.NET 7.0` for testing

File: framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs (hunk 2 above)
Comment (sunnamed434) [reply framework/OpenMod.Core/Localization/ProxyStringLocalizer.cs:30]: Done

File: framework/OpenMod.Core/Localization/NullStringLocalizer.cs (hunk 1 above)
Comment (sunnamed434) [reply framework/OpenMod.Core/Localization/NullStringLocalizer.cs:30]: Done

File: rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj (hunk 9 above)
[the remark this answers is not in the document]
Comment (sunnamed434) [reply rust/oxide/tests/OpenMod.Rust.Oxide.Tests/OpenMod.Rust.Oxide.Tests.csproj]: Done

**Review by DiFFoZ — APPROVED** [review]
Tested those plugins and they work fine:
```yaml
- id: OpenMod.Economy
  version: 1.3.12
- id: DiFFoZ.Kits
  version: 2.4.2
- id: Pustalorc.PlayerInfoLib.Unturned
  version: 3.6.6
- id: Wild.Announcer
  version: 1.1.3
- id: Kr4ken.NewEssentials
  version: 1.5.1
- id: EvolutionPlugins.Dummy
  version: 2.7.0
- id: EvolutionPlugins.OpenDeathMessages
  version: 1.1.1
```

But still requires more testing.
###### OpenMod nightly builds when 🥹

**Comment by DiFFoZ** [comment]
Currently blocking to make decision about updating Harmony to the [Harmony v2.3-prerelease.2](https://github.com/pardeike/Harmony/releases/tag/v2.3-prerelease.2), because v2.2.2 causes compile issue with transpiler patch (`Label` reference mismatch, see image).
image

**Comment by sunnamed434** [comment]
> Currently blocking to make decision about updating Harmony to the [Harmony v2.3-prerelease.2](https://github.com/pardeike/Harmony/releases/tag/v2.3-prerelease.2), because v2.2.2 causes compile issue with transpiler patch (`Label` reference mismatch, see image). image

Perhaps it is more a VS problem, I use JetBrains Rider, and it is okay for me, but anyway, I changed the version to the latest Harmony pre-release v3, updated runtime libs, and fixed other issues

File: unturned/OpenMod.Unturned.Module.Dev/runtimelibs/0Harmony.dll
Comment (DiFFoZ) [inline unturned/OpenMod.Unturned.Module.Dev/runtimelibs/0Harmony.dll:1]: Why Harmony is added as a file, instead of package reference?

File: unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs [hunk 11]
```diff
@@ -1,7 +1,8 @@
-﻿using SDG.Unturned;
+﻿extern alias MonoModILHelpers;
+using MonoModILHelpers::System.Runtime.CompilerServices;
```
Comment (DiFFoZ) [inline unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs:2]: Don't use `extern alias` of monomod, use the .net standard API

**Review by DiFFoZ — CHANGES_REQUESTED** [review]

File: unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs (hunk 11 above)
Comment (sunnamed434) [reply unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs:2]: Try to remove it, you will get an error due to multiple Unsafe types with the same name and namespace

Comment (sunnamed434) [reply unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs:2]: If you will try to change alias name to system (i.e netstandard) it won't work, idk why

File: unturned/OpenMod.Unturned.Module.Dev/runtimelibs/0Harmony.dll
Comment (sunnamed434) [reply unturned/OpenMod.Unturned.Module.Dev/runtimelibs/0Harmony.dll:1]: Removed 0Harmony dlls from runtime libs

File: unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs (hunk 11 above)
Comment (sunnamed434) [reply unturned/OpenMod.Unturned/Events/UnturnedPlayerEventsListener.cs]: Fixed

File: framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs [hunk 12]
```diff
@@ -41,41 +42,55 @@ private async Task<int> GetPageNumberAsync()
...
-            await PrintAsync($"[{pageNumber}/{pageCount}] OpenMod Plugins", Color.CornflowerBlue);
+            var sb = new StringBuilder();
+
+            sb.Append('[');
+            sb.Append(pageNumber);
+            sb.Append('/');
+            sb.Append(pageCount);
+            sb.Append("] OpenMod Plugins");
+
+            await PrintAsync(sb.ToString(), Color.CornflowerBlue);
```
Comment (Trojaner) [inline framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs:69]: ?

File: framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs [hunk 13]
```diff
@@ -41,41 +42,55 @@ private async Task<int> GetPageNumberAsync()
...
+            sb.Append(plugin.Version);
 
             if (!string.IsNullOrEmpty(plugin.Author))
             {
-                sb.Append($" by {plugin.Author}");
+                sb.Append(" by ");
+                sb.Append(plugin.Author);
             }
 
             return PrintAsync(sb.ToString(), Color.Green);
```
Comment (Trojaner) [inline framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs:96]: ?

Comment (sunnamed434) [reply framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs]: Fixed

File: framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs (hunk 12 above)
Comment (sunnamed434) [reply framework/OpenMod.Core/Commands/OpenModCommands/CommandOpenModPlugins.cs]: Fixed

**Comment by sunnamed434** [comment]
Fixed branch conflicts

**Comment by sunnamed434** [comment]
Fixed conflicts

**Comment by sunnamed434** [comment]
Fixed conflicts

**Comment by sunnamed434** [comment]
Fixed conflicts

**Comment by Trojaner** [comment]
This will probably be merged with OpenMod 3.8

**Comment by DiFFoZ** [comment]
Before releasing the new version, we should ping devs about the EF Core error that currently is forced to silent, see commit https://github.com/openmod/openmod/pull/772/commits/6e73e4d7bcb155c0e23784cdc90a1251a656256f

**Comment by Trojaner** [comment]
Given the huge amount of conflicts and age, I redid the PR and merged manually.
Thanks for your contributions.
